src/rust/engine/src/nodes.rs

```diff
@@ -747,62 +749,66 @@ impl DownloadedFile {
             url,
           ))
         } else {
-          Ok((file_name, response))
+          Ok(response)
+        }
+      }).and_then(move |response| {
+        struct SizeLimiter<W: std::io::Write> {
+          writer: W,
+          written: usize,
+          size_limit: usize,
+        }
+
+        impl<W: std::io::Write> Write for SizeLimiter<W> {
+          fn write(&mut self, buf: &[u8]) -> Result<usize, std::io::Error> {
+            let new_size = self.written + buf.len();
+            if new_size > self.size_limit {
+              Err(std::io::Error::new(
+                std::io::ErrorKind::InvalidData,
+                "Downloaded file was larger than expected digest",
+              ))
+            } else {
+              self.written = new_size;
+              self.writer.write_all(buf)?;
+              Ok(buf.len())
+            }
+          }
+
+          fn flush(&mut self) -> Result<(), std::io::Error> {
+            self.writer.flush()
+          }
         }
-      }).and_then(move |(file_name, response)| {
-        let tempdir = try_future!(
-          tempfile::TempDir::new()
-            .map_err(|err| format!("Error making tempdir to download file: {}", err))
-        );
-
-        // Copy the contents of response into a file and modify its permissions.
-        let temp_file_path = tempdir.path().join(&file_name);
-        future::done(std::fs::File::create(&temp_file_path))
-          .and_then(|file| {
-            use std::os::unix::fs::PermissionsExt;
-            std::fs::metadata(temp_file_path).and_then(|metadata| {
-              let mut permissions = metadata.permissions();
-              permissions.set_mode(0o755);
-              file.set_permissions(permissions)?;
-              Ok(file)
-            })
-          }).map_err(|err| format!("Error making temporary file for URL fetch: {}", err))
-          .and_then(|file| {
-            let hasher = hashing::WriterHasher::new(file);
-            response
-              .into_body()
-              .map_err(|err| format!("Error reading URL fetch response: {}", err))
-              .fold(hasher, |mut hasher, chunk| {
-                hasher
-                  .write_all(&chunk)
-                  .map(|_| hasher)
-                  .map_err(|err| format!("Error hashing/writing URL fetch response: {}", err))
-              }).map(|hasher| (hasher.finish(), file_name, tempdir))
-          }).to_boxed()
-      }).and_then(move |(actual_digest, file_name, tempdir)| {
+
+        let hasher = hashing::WriterHasher::new(SizeLimiter {
```

Comment:
You found an instance where I miss the Java anonymous classes to implement interfaces like this. That said, I don't disagree with this code.